src/vendor_app/app/src/main/kotlin/cz/quanti/android/vendor_app/repository/purchase/impl/PurchaseFacadeImpl.kt

```diff
@@ -0,0 +1,68 @@
+package cz.quanti.android.vendor_app.repository.purchase.impl
+
+import cz.quanti.android.vendor_app.repository.card.CardRepository
+import cz.quanti.android.vendor_app.repository.purchase.PurchaseFacade
+import cz.quanti.android.vendor_app.repository.purchase.PurchaseRepository
+import cz.quanti.android.vendor_app.repository.purchase.dto.Purchase
+import cz.quanti.android.vendor_app.utils.BlockedCardError
+import cz.quanti.android.vendor_app.utils.VendorAppException
+import cz.quanti.android.vendor_app.utils.isPositiveResponseHttpCode
+import io.reactivex.Completable
+import io.reactivex.Observable
+
+class PurchaseFacadeImpl(
+    private val purchaseRepo: PurchaseRepository,
+    private val cardRepo: CardRepository
+) : PurchaseFacade {
+
+    override fun savePurchase(purchase: Purchase): Completable {
+        if (purchase.smartcard != null) {
+            return cardRepo.getBlockedCard(purchase.smartcard!!).flatMapCompletable { itsBlocked ->
+                throw BlockedCardError("This card is tagged as blocked on the server")
+            }.andThen(purchaseRepo.savePurchase(purchase))
+        } else {
+            return purchaseRepo.savePurchase(purchase)
+        }
+    }
+
+    override fun syncWithServer(): Completable {
+        return sendPurchasesToServer()
+            .andThen(deleteAllPurchases())
+    }
+
+    private fun sendPurchasesToServer(): Completable {
+        return purchaseRepo.getAllPurchases().flatMapCompletable { purchases ->
+            purchaseRepo.sendVoucherPurchasesToServer(purchases)
+                .defaultIfEmpty(200)
```

Comment:
Ruhle lišárnu bych asi udělal obráceně. sendVoucherPurchasesToServer by neřešila prázdnej list, to bys řešila tady